### PR TITLE
[data][train] Update "Data Loading and Preprocessing" doc

### PR DESCRIPTION
**Pull request by raulchen**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 1]
```diff
@@ -3,12 +3,13 @@
 Data Loading and Preprocessing
 ==============================
 
-Ray Train integrates with :ref:`Ray Data <data>` to offer an efficient, streaming solution for loading and preprocessing large datasets.
-Use Ray Data to performantly run large-scale distributed training workloads. Key advantages include:
+Ray Train integrates with :ref:`Ray Data <data>` to offer an performant and scalable streaming solution for loading and preprocessing large datasets.
```
Comment (scottjlee) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:6]:
```suggestion
Ray Train integrates with :ref:`Ray Data <data>` to offer a performant and scalable streaming solution for loading and preprocessing large datasets.
```

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 2]
```diff
@@ -3,12 +3,13 @@
...
+Key advantages include:
 
-- Fast out-of-memory recovery
-- Support for heterogeneous clusters
-- No dropped rows during distributed dataset iteration
+- Streaming data loading and preprocessing, scalable to petabyte-scale data.
+- Scaling out heavy data preprocessing to CPU nodes and avoid GPU training being bottlenecked.
```
Comment (scottjlee) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:10]:
```suggestion
- Scaling out heavy data preprocessing to CPU nodes, avoiding GPU training from being bottlenecked.
```

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 3]
```diff
@@ -30,17 +31,17 @@ Install Ray Data and Ray Train:
 
 Data ingestion can be set up with four basic steps:
 
-1. Create a Ray Dataset.
-2. Preprocess your Ray Dataset.
-3. Input the preprocessed Dataset into the Ray Train Trainer.
+1. Create a Ray Dataset from your input data.
+2. Apply preprocessing operations to your Ray Dataset.
+3. Input the preprocessed Dataset into the Ray Train Trainer, which internally will split the dataset equally in a streaming way across the distributed training workers.
```
Comment (scottjlee) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:36]:
```suggestion
3. Input the preprocessed Dataset into the Ray Train Trainer, which internally splits the dataset equally in a streaming way across the distributed training workers.
```

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 4]
```diff
@@ -213,6 +223,8 @@ The datasets passed into the Trainer's ``datasets`` can be accessed inside of th
 
 All datasets are split (i.e. sharded) across the training workers by default. :meth:`~ray.train.get_dataset_shard` will return ``1/n`` of the dataset, where ``n`` is the number of training workers.
 
+Data splitting is done in a streaming fashion on the fly.
```
Comment (scottjlee) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:226]: should we combine this into the previous paragraph?

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 5]
```diff
@@ -416,14 +428,34 @@ Random shuffling
 ----------------
 Randomly shuffling data for each epoch can be important for model quality depending on what model you are training.
 
-Ray Data has two approaches to random shuffling:
+Ray Data provides following options to random shuffling, sorted by compute overheads and randomness in acceding order. These options can be used in combination.
```
Comment (scottjlee) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:431]:
```suggestion
Ray Data provides the following options for random shuffling, sorted by compute overheads and randomness in acceding order. These options can be used in combination.
```

Comment (scottjlee) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:431]: by the way, i added this content in a new page under Ray Data: https://docs.ray.io/en/master/data/shuffling-data.html#types-of-shuffling

we could either link to that page, or omit this section.

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 6]
```diff
@@ -565,14 +616,14 @@ You can use this with Ray Train Trainers by applying them on the dataset before
...
+In this example, we persist the fitted preprocessor using the ``Trainer(metadata={...})`` constructor argument. This arg specifies a dict that will be available from ``TrainContext.get_metadata()`` and ``checkpoint.get_metadata()`` for checkpoints saved from the Trainer. This enables recreation of the fitted preprocessor for use for inference.
 
 Performance tips
 ----------------
 
 Prefetching batches
 ~~~~~~~~~~~~~~~~~~~
-While iterating over your dataset for training, you can increase ``prefetch_batches`` in :meth:`iter_batches <ray.data.DataIterator.iter_batches>` or :meth:`iter_torch_batches <ray.data.DataIterator.iter_torch_batches>` to further increase performance. While training on the current batch, this launches N background threads to fetch and process the next N batches.
+While iterating over your dataset for training, you can increase ``prefetch_batches`` in :meth:`iter_batches <ray.data.DataIterator.iter_batches>` or :meth:`iter_torch_batches <ray.data.DataIterator.iter_torch_batches>` to further increase performance. While training on the current batch, this launches background threads to fetch and process the next N batches.
```
Comment (scottjlee) [inline doc/source/train/user-guides/data-loading-preprocessing.rst:626]:
```suggestion
While iterating over your dataset for training, you can increase ``prefetch_batches`` in :meth:`iter_batches <ray.data.DataIterator.iter_batches>` or :meth:`iter_torch_batches <ray.data.DataIterator.iter_torch_batches>` to further increase performance. While training on the current batch, this launches background threads to fetch and process the next ``N`` batches.
```

**Review by scottjlee — APPROVED** [review]
pending fixing docs/premerge failures

**Review by angelinalg — APPROVED** [review]
Please correct any subjects that I added to convert passive voice sentences to active voice.

Also, consider using Vale to catch spelling and style errors. (go/vale)

File: doc/source/train/user-guides/data-loading-preprocessing.rst (hunk 5 above)
Comment (raulchen) [reply doc/source/train/user-guides/data-loading-preprocessing.rst:431]: will link to that page

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 7]
```diff
@@ -71,6 +72,9 @@ Data ingestion can be set up with four basic steps:
                 # Step 4: Access the dataset shard for the training worker via
```
[the remark this answers is not in the document]
Comment (raulchen) [reply doc/source/train/user-guides/data-loading-preprocessing.rst:72]: from runtime's perspective, this happens after the step 3 below

File: doc/source/train/user-guides/data-loading-preprocessing.rst [hunk 8]
```diff
@@ -54,7 +55,7 @@ Data ingestion can be set up with four basic steps:
 
             # Step 1: Create a Ray Dataset from in-memory Python lists.
             # You can also create a Ray Dataset from many other sources and file
-            # formats.
+            # formats, see https://docs.ray.io/en/latest/data/loading-data.html.
```
[the remark this answers is not in the document]
Comment (raulchen) [reply doc/source/train/user-guides/data-loading-preprocessing.rst]: removed this change.
